src/libcore/str/pattern.rs

```diff
@@ -582,9 +594,22 @@ impl<'a, 'b> StrSearcher<'a, 'b> {
 }
 
 unsafe impl<'a, 'b> Searcher<'a> for StrSearcher<'a, 'b> {
+    #[inline]
     fn haystack(&self) -> &'a str { self.haystack }
 
     #[inline]
+    fn remaining_haystack(&self) -> &'a str {
+        match self.searcher {
+            StrSearcherImpl::Empty(EmptyNeedle { position, end, .. }) |
+            StrSearcherImpl::TwoWay(TwoWaySearcher { position, end, .. }) => {
+                unsafe {
+                    self.haystack.slice_unchecked(position, end)
```

Comment:
But I would like to not do this, it's another restriction on the string search algorithm (that it can seek to an appropriate place to cut the remaining part, and that it hasn't jumped far ahead for some reason).